### ISSUE TITLE
Run WebVR content without context switching?

### ISSUE THREAD

**Issue by asajeffrey** [servo/servo#22855]
At the moment, running WebVR content keeps active:

- The main thread
- At least one script thread
- At least one layout thread (even though we never render the resulting laid out DOM)
- The constellation
- The webGL thread
- The webVR thread (and possibly the webVR rAF thread)
- The webrender thread (though that could be merged with the main thread?)
- Forwarding threads copying messages from IPC channels to MPSC channels

According to https://www.magicleap.com/magic-leap-one there are three cores available for magic leap applications, so with the current architecture we are going to be doing quite a bit of context switching. Reducing the context switching would require merging or eliminating some of the threads.

**Comment by MortimerGoro** [comment]
On FirefoxReality using GeckoView we pause the Gecko compositor when a WebVR session starts. I think some of the Servo threads could also be stopped if we can do something similar as pausing the compositor (I mean not waking up for processing events until we enable it again)

> The main thread

Probably a different sceneario between Magic Leap and other. On Oculus or Daydream we have a more flexible way to run the VR calls in a custom thread.

> At least one script thread

No way to stop this because of the JavaScript code

> At least one layout thread (even though we never render the resulting laid out DOM)

A WebVR session shouldn't require DOM layout or repaint. We could bypass this rendering to a WebGL canvas directly. DOMToTexture or Quad Layers may change this on the future once it's standarized.

> * The constellation

A WebVR session render loop uses WebGL/WebVR calls. They don't normally need to awake the constellation.

> * The webGL thread

This is required for all the drawing. Currently also WebVR compositor calls (e.g.  `SubmitFrame`
and `sync_poses`) are run in this thread. 

> The webVR thread (and possibly the webVR rAF thread)

Currently on servo the WebVR thread is not normally waken up during a WebVR session (unless some event is dispatched). The Canvas output is send directly to the WebGL thread via a direct channel. The most important calls in a WebVR session,  `getFrameData` and `submitFrame` directly use a channel between script thread and WebGL thread.

> * The webrender thread (though that could be merged with the main thread?)

If we can do something similar to pausing the compositor webrender thread could be inactive.

> * Forwarding threads copying messages from IPC channels to MPSC channels

The channel used for `getFrameData` and `submitFrame` is direct between script thread and WebGL thread. So no forwarding threads should be doing work during a WebVR session for those calls.

So I think that ideally we could reduce that list to:

* The main thread (On Magic Leap, don't required for e.g. Oculus)
* Script thread
* WebGL thread
* WebVR RAF thread (which maybe can be merged with WebVR thread in the future, once we have https://github.com/servo/rust-offscreen-rendering-context/issues/133 implemented)

**Comment by asajeffrey** [comment]
> On FirefoxReality using GeckoView we pause the Gecko compositor when a WebVR session starts. I think some of the Servo threads could also be stopped if we can do something similar as pausing the compositor (I mean not waking up for processing events until we enable it again)

Yes, we could try pausing style/layout/WR, though as long as the content doesn't update the DOM or run any animations on the DOM content, those threads aren't doing much.

> > The main thread
> 
> Probably a different sceneario between Magic Leap and other. On Oculus or Daydream we have a more flexible way to run the VR calls in a custom thread.

Yeah, I can imagine it's different on different devices.

> > At least one script thread
> 
> No way to stop this because of the JavaScript code
> 
> > At least one layout thread (even though we never render the resulting laid out DOM)
> 
> A WebVR session shouldn't require DOM layout or repaint. We could bypass this rendering to a WebGL canvas directly. DOMToTexture or Quad Layers may change this on the future once it's standarized.

Rather annoyingly, the content is allowed to update the DOM then query it, so if the content updates the DOM we need to run layout. But for pages that don't update the DOM page we're okay.

> > * The constellation
> 
> A WebVR session render loop uses WebGL/WebVR calls. They don't normally need to awake the constellation.

Keyboard and mouse events go via the constellation, so on devices which generate such events, they're going to wake up the constellation.

> > * The webGL thread
> 
> This is required for all the drawing. Currently also WebVR compositor calls (e.g. `SubmitFrame`
> and `sync_poses`) are run in this thread.
> 
> > The webVR thread (and possibly the webVR rAF thread)
> 
> Currently on servo the WebVR thread is not normally waken up during a WebVR session (unless some event is dispatched). The Canvas output is send directly to the WebGL thread via a direct channel. The most important calls in a WebVR session, `getFrameData` and `submitFrame` directly use a channel between script thread and WebGL thread.

Does `schedule_poll_events` not cause the poll events thread to be spawned off and wake up the WebVR thread when polling?

> > * The webrender thread (though that could be merged with the main thread?)
> 
> If we can do something similar to pausing the compositor webrender thread could be inactive.

True. It's only active when the content updates the DOM.

> > * Forwarding threads copying messages from IPC channels to MPSC channels
> 
> The channel used for `getFrameData` and `submitFrame` is direct between script thread and WebGL thread. So no forwarding threads should be doing work during a WebVR session for those calls.

True as long as we can get the constellation out of the way, otherwise that uses a `ROUTER`, which is another thread.

> So I think that ideally we could reduce that list to:
> 
>     * The main thread (On Magic Leap, don't required for e.g. Oculus)
> 
>     * Script thread
> 
>     * WebGL thread
> 
>     * WebVR RAF thread (which maybe can be merged with WebVR thread in the future, once we have [servo/rust-offscreen-rendering-context#133](https://github.com/servo/rust-offscreen-rendering-context/issues/133) implemented)

As long as the content doesn't update the DOM, there's no keyboard/mouse events, and `schedule_poll_events` doesn't cause polling, this sounds right. Four threads isn't so bad, on much hardware it's less than the available cores.

**Comment by MortimerGoro** [comment]
`Does `schedule_poll_events` not cause the poll events thread to be spawned off and wake up the WebVR thread when polling?`

True, but the event polling interval was only 500ms, so most of the time is inactive.

We could probably optimize it to only wake up when receiving a notification, or move the event check to the WebVR RAF thread

**Comment by asajeffrey** [comment]
Yeah, it would be nice to do something about that polling, I can imagine it causing dropped frames every 500ms, which isn't ideal.

**Comment by pcwalton** [comment]
I think that if the `ROUTER` thread is a problem, then we should fix that somehow instead of adding anything WebXR-specific.
